python2/typing.py

```diff
@@ -1081,6 +1065,38 @@ def __init__(self, *args, **kwargs):
         super(GenericMeta, self).__init__(*args, **kwargs)
         if isinstance(self.__extra__, abc.ABCMeta):
             self._abc_registry = self.__extra__._abc_registry
+            self._abc_cache = self.__extra__._abc_cache
+        elif self.__origin__ is not None:
+            self._abc_registry = self.__origin__._abc_registry
+            self._abc_cache = self.__origin__._abc_cache
```

Comment:
``self._abc_negative_cache`` is also shared, but it could be overwritten by ``abc.py`` on negative cache invalidation, this is why I made it a descriptor 7 lines below. (Ppositive cache is never invalidated, so that the sharing could be realised as a simple assignment on instantiation).